### PR TITLE
feat(discord): Implement Discord identity linking/unlinking

### LINKED ISSUE #50894: Implement Discord identity linking and unlinking via slash command
- [x] Add commands to `IntegrationProvider.setup()`
- [x] Implement handler (see `Identity` model and how Slack does this)
- [x] Make sure handler is called when we receive the relevant events at our interactions endpoint

### PR DESCRIPTION
**Pull request by spalmurray**
Create Discord Identity provider. Identity for installing user will be created on integration installation by completing the OAuth2 flow to get this user's ID. After this, other Discord users will be able to link their identities through the /link command, and unlink through the /unlink command. No OAuth2 flow is needed for users after the first, we will just link their Discord user id to their Sentry user.

### CODE REVIEW

File: src/sentry/identity/discord/provider.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from sentry.auth.providers.oauth2 import OAuth2Provider
+
+
+class DiscordIdentityProvider(OAuth2Provider):
+    key = "discord"
+    name = "Discord"
+
+    # We don't actually use OAuth2 for Discord identities, so this class is
```
Comment (evanpurkhiser) [inline src/sentry/identity/discord/provider.py:8]: Does it need to extend `OAuth2Provider` then? Can it just extend `Provider`?

Comment (spalmurray) [reply src/sentry/identity/discord/provider.py:8]: Yes it can! Thanks for the suggestion @evanpurkhiser 😄

Comment (evanpurkhiser) [reply src/sentry/identity/discord/provider.py:8]: Why doesn't it use OAuth2 btw? What procedure does it use for identity linking?

Comment (spalmurray) [reply src/sentry/identity/discord/provider.py:8]: My thinking is that we only need their Discord user id. We tie that to Sentry user via the Identity and when that Discord user takes actions on a notification in Discord, we'll know who's doing it. We don't need to go through the OAuth2 flow to get this user ID, Discord gives it to us when they use the `/link` command. 

Lmk if my thinking is flawed or there's some other reason to do the whole flow.

Comment (evanpurkhiser) [reply src/sentry/identity/discord/provider.py:8]: I don't know enough about how this integration works to be able to advise on that I think.

Would be happy to sit down and walk through it all some time if you'd like though!

File: src/sentry/integrations/discord/commands.py [hunk 2]
```diff
@@ -14,8 +14,18 @@ class DiscordCommandManager:
...
+            "type": 1,
+        },
+        {
+            "name": "unlink",
+            "description": "Unlink your Discord account from your Sentry account.",
+            "type": 1,
+        },
+        {
+            "name": "help",
+            "description": "View a list of Sentry bot commands and what they do",
```
Comment (snigdhas) [inline src/sentry/integrations/discord/commands.py:28]: Let's be consistent on whether we do/don't use punctuation

File: src/sentry/integrations/discord/integration.py [hunk 3]
```diff
@@ -16,6 +17,7 @@
 from sentry.integrations.discord.client import DiscordClient
 from sentry.integrations.discord.commands import DiscordCommandManager
 from sentry.pipeline.views.base import PipelineView
+from sentry.shared_integrations.exceptions import IntegrationError
 from sentry.shared_integrations.exceptions.base import ApiError
 from sentry.utils.http import absolute_uri
```
Comment (snigdhas) [inline src/sentry/integrations/discord/integration.py:23]: didn't catch this earlier but we usually use

`from logging import logger` 

followed by 

`logger = logging.getLogger(__name__)` to report logs associated with the right path/package.

File: src/sentry/integrations/discord/integration.py [hunk 4]
```diff
@@ -108,40 +110,85 @@ def __init__(self) -> None:
...
+            token = response["access_token"]  # type: ignore
+
+            headers = {"Authorization": f"Bearer {token}"}
+
+            # Can't use self.client.get because that will overwrite our headers
+            # with our bot's authorization
+            response = requests.get(f"{self.client.base_url}/users/@me", headers=headers)
+            return response.json()["id"]
+        except (ApiError, KeyError):
+            logger.error("discord.install.failed_to_get_discord_user_id")
```
Comment (snigdhas) [inline src/sentry/integrations/discord/integration.py:184]: Do we want to include any additional info about the error/request? you can include a dict `extra` like so:

```
logger.error(
            "discord.install.failed_to_get_discord_user_id",
            extra={
                "key": value
            },
        )
```

**Review by snigdhas — APPROVED** [review]

**Pull request merged**